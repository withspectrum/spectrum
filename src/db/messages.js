import * as firebase from 'firebase';
import { getStory } from './stories';
import { createNotifications } from './notifications';
import { ACTIVITY_TYPES, OBJECT_TYPES } from './types';

/**
 * Create a message in the db
 *
 * Resolves the returned promise with the created message data
 */
export const createMessage = ({ storyId, frequency, userId, message }) => {
  const db = firebase.database();

  const key = db.ref('messages').push().key;

  return db
    .ref()
    .update({
      [`messages/${key}`]: {
        id: key,
        storyId,
        frequencyId: frequency.id,
        timestamp: firebase.database.ServerValue.TIMESTAMP,
        userId,
        message,
      },
      [`stories/${storyId}/messages/${key}`]: {
        id: key,
      },
    })
    .then(() => db.ref(`messages/${key}`).once('value'))
    .then(snapshot => {
      const data = snapshot.val();
      createNotifications({
        // Only add notifications for other users
        users: Object.keys(frequency.users).filter(user => user !== userId),
        activityType: ACTIVITY_TYPES.NEW_MESSAGE,
        objectType: OBJECT_TYPES.STORY,
        objectId: storyId,
        objectUrl: `https://spectrum.chat/~${frequency.slug ||
          frequency.id}/${storyId}`,
        senderId: userId,
      });
      return data;
    });
};

export const getMessage = messageId => {
  const db = firebase.database();

  return db
    .ref(`messages/${messageId}`)
    .once('value')
    .then(snapshot => snapshot.val());
};

export const getMessages = storyId => {
  return getStory(storyId).then(story => {
    if (!story.messages) return Promise.resolve([]);
    const messages = Object.keys(story.messages);
    return Promise.all(messages.map(message => getMessage(message)));
  });
};
