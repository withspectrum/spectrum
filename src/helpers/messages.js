// @flow
import { sortByDate } from './utils';

// TODO: FIXME
type Message = Object;

export const sortAndGroupMessages = (
  messages: Array<Message>,
  lastSeen?: number
) => {
  if (messages.length === 0) return [];
  messages = sortByDate(messages, 'timestamp', 'asc');
  let masterArray = [];
  let newArray = [];
  let hasInjectedUnseenRobo = false;
  let checkId;

  const unseenRobo = [
    {
      author: {
        user: {
          id: 'robo',
        },
      },
      timestamp: lastSeen,
      message: {
        content: '',
        type: 'unseen-messages-below',
      },
    },
  ];

  for (let i = 0; i < messages.length; i++) {
    // on the first message, get the user id and set it to be checked against
    const robo = [
      {
        author: {
          user: {
            id: 'robo',
          },
        },
        timestamp: messages[i].timestamp,
        message: {
          content: messages[i].timestamp,
          type: 'timestamp',
        },
      },
    ];

    // If the message is an optimistic response that means the user's seen all messages
    // so we remove any already injected lastSeen robo and make it not inject any more
    if (typeof messages[i].id === 'number') {
      if (hasInjectedUnseenRobo) {
        masterArray = masterArray.filter(group => group !== unseenRobo);
      }

      hasInjectedUnseenRobo = true;
    }

    if (i === 0) {
      checkId = messages[i].author.user.id;

      if (messages[0].timestamp > lastSeen && !hasInjectedUnseenRobo) {
        hasInjectedUnseenRobo = true;
        masterArray.push(unseenRobo);
      } else {
        masterArray.push(robo);
      }
    }

    const sameUser =
      messages[i].author.user.id !== 'robo' &&
      messages[i].author.user.id === checkId; //=> boolean
    const oldMessage = (current: Object, previous: Object) => {
      //=> boolean
      /*
        Rethink db returns string timestamps. We need to convert them
        into a number format so that we can compare message timestamps
        more easily. .getTime() will convert something like:

        '2017-05-02T18:15:20.769Z'

        into:

        '1493748920.769'

        We then determine if the current message being evaluated is older
        than the previous evaulated message by a certain integer to determine
        if we should render a timestamp in the UI and create a new bubbleGroup
      */
      const c = new Date(current.timestamp).getTime();
      const p = new Date(previous.timestamp).getTime();
      return c > p + 21600000;
    };

    // if we are evaulating a bubble from the same user
    if (sameUser) {
      // if we are still on the first message
      if (i === 0) {
        // push the message to the array
        newArray.push(messages[i]);
      } else {
        if (messages[i].timestamp > lastSeen && !hasInjectedUnseenRobo) {
          hasInjectedUnseenRobo = true;
          masterArray.push(newArray);
          masterArray.push(unseenRobo);
          newArray = [];
        }
        // if we're on to the second message, we need to evaulate the timestamp
        // if the second message is older than the first message by our variance
        if (oldMessage(messages[i], messages[i - 1])) {
          // push the batch of messages to master array
          masterArray.push(newArray);
          // insert a new robotext timestamp
          masterArray.push(robo);
          // reset the batch of new messages
          newArray = [];
          // populate the new batch of messages with this next old message
          newArray.push(messages[i]);
        } else {
          // if the message isn't older than our prefered variance,
          // we keep populating the same batch of messages
          newArray.push(messages[i]);
        }
      }
      // and maintain the checkid
      checkId = messages[i].author.user.id;
      // if the next message is from a new user
    } else {
      // we push the previous user's messages to the masterarray
      masterArray.push(newArray);
      if (messages[i].timestamp > lastSeen && !hasInjectedUnseenRobo) {
        hasInjectedUnseenRobo = true;
        masterArray.push(unseenRobo);
      }
      // if the new users message is older than our preferred variance
      if (i > 0 && oldMessage(messages[i], messages[i - 1])) {
        // push a robo timestamp
        masterArray.push(robo);
        newArray = [];
        newArray.push(messages[i]);
      } else {
        // clear the messages array from the previous user
        newArray = [];
        // and start a new batch of messages from the currently evaulating user
        newArray.push(messages[i]);
      }

      // set a new checkid for the next user
      checkId = messages[i].author.user.id;
    }
  }

  // when done, push the final batch of messages to masterArray
  // masterArray.push(newArray);
  // and return masterArray to the component
  masterArray.push(newArray);
  return masterArray;
};
