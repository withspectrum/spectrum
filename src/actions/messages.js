import { createMessage } from '../db/messages';
import { track } from '../EventTracker';
import { getCurrentFrequency } from '../helpers/frequencies';

/**
 * Send a message
 */
export const sendMessage = message => (dispatch, getState) => {
  const { user, stories, frequencies: { frequencies } } = getState();
  const storyId = stories.active;
  const frequencyId = stories.stories.find(
    story => story.id === stories.active,
  ).frequencyId;
  const frequency = getCurrentFrequency(frequencyId, frequencies);

  track('message', 'sent', null);

  createMessage({
    storyId,
    frequency,
    userId: user.uid,
    message,
  });
};
