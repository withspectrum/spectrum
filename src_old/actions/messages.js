import {
  createMessage,
  createPrivateMessage,
  setLastActivityOnMessageGroupParticipants,
  getMessageKey,
  createReaction,
  deleteReaction,
} from '../db/messages';
import { createNewMessageGroup } from '../db/messageGroups';
import { track } from '../EventTracker';
import { getCurrentFrequency } from '../helpers/frequencies';
import history from '../helpers/history';
import { throwError } from './errors';

/**
 * Send a message
 */
export const sendMessage = message => (dispatch, getState) => {
  const {
    user,
    stories,
    frequencies: { frequencies },
    messageGroups,
    messageComposer,
  } = getState();
  const storyId = stories.active;
  const messageGroupId = messageGroups.active;

  if (storyId) {
    // if the user is sending this message from a story
    const frequencyId = stories.stories.find(
      story => story.id === stories.active
    ).frequencyId;
    const frequency = getCurrentFrequency(frequencyId, frequencies);

    track(`${message.type} message`, 'sent', null);

    const key = getMessageKey('messages');

    // Show message locally before persisting to server
    dispatch({
      type: 'SEND_MESSAGE',
      message: {
        userId: user.uid,
        timestamp: Date.now(),
        frequencyId: frequency.id,
        storyId: storyId,
        id: key,
        message,
        persisted: false,
      },
    });

    createMessage({
      storyId,
      frequencyId: frequency.id,
      user,
      message,
      key,
    });
  }

  if (messageGroupId) {
    // if the user is in a private message group
    const messageGroup = messageGroups.messageGroups.find(
      messageGroup => messageGroup.id === messageGroupId
    );

    track(`messageGroup ${message.type} message`, 'sent', null);

    const key = getMessageKey('messages_private');

    // Show message locally before persisting to server
    dispatch({
      type: 'SEND_MESSAGE',
      message: {
        userId: user.uid,
        timestamp: Date.now(),
        messageGroupId: messageGroup.id,
        id: key,
        message,
        persisted: false,
      },
    });

    createPrivateMessage({
      messageGroupId,
      userId: user.uid,
      message,
      key,
    }).then(() => {
      setLastActivityOnMessageGroupParticipants(messageGroup);
    });
  }

  if (!storyId && !messageGroupId) {
    // if there is neither a story or message, assume the user is messaging another
    // user for the first time
    const recipient = messageComposer.recipient.uid;
    createNewMessageGroup(user.uid, recipient).then(newGroupKey => {
      track(`messageGroup ${message.type} message`, 'sent', null);

      const key = getMessageKey('messages_private');

      // Show message locally before persisting to server
      dispatch({
        type: 'SEND_MESSAGE',
        message: {
          userId: user.uid,
          timestamp: Date.now(),
          messageGroupId: newGroupKey,
          id: key,
          message,
          persisted: false,
        },
      });

      createPrivateMessage({
        messageGroupId: newGroupKey,
        userId: user.uid,
        message,
        key,
      }).then(() => {
        history.push(`/messages/${newGroupKey}`);

        dispatch({
          type: 'SET_ACTIVE_MESSAGE_GROUP',
          messageGroupId: newGroupKey,
        });

        dispatch({
          type: 'CLOSE_MESSAGE_COMPOSER',
        });
      });
    });
  }
};

export const addReaction = messageId => (dispatch, getState) => {
  const { user: { uid }, stories, messageGroups } = getState();
  const activeStory = stories.active;
  const activeMessageGroup = messageGroups.active;
  const type = activeStory
    ? 'messages'
    : activeMessageGroup ? 'messages_private' : 'messages';

  track('reaction', 'created', null);

  createReaction({
    type,
    messageId,
    uid,
  })
    .then(() => {
      dispatch({
        type: 'ADD_REACTION',
        messageId,
        uid,
      });
    })
    .catch(err => {
      dispatch(throwError(err));
    });
};

export const removeReaction = messageId => (dispatch, getState) => {
  const { user: { uid }, stories, messageGroups } = getState();
  const activeStory = stories.active;
  const activeMessageGroup = messageGroups.active;
  const type = activeStory
    ? 'messages'
    : activeMessageGroup ? 'messages_private' : 'messages';

  track('reaction', 'removed', null);

  deleteReaction({
    type,
    messageId,
    uid,
  })
    .then(() => {
      dispatch({
        type: 'REMOVE_REACTION',
        messageId,
        uid,
      });
    })
    .catch(err => {
      dispatch(throwError(err));
    });
};
