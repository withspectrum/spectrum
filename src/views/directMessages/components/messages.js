// @flow
import React, { Component } from 'react';
// $FlowFixMe
import compose from 'recompose/compose';
// $FlowFixMe
import pure from 'recompose/pure';
import { sortAndGroupMessages } from '../../../helpers/messages';
import ChatMessages from '../../../components/chatMessages';
import { Loading } from '../../../components/loading';
import { Spinner } from '../../../components/globals';
import { getDirectMessageThreadMessages } from '../queries';
import { toggleReactionMutation } from '../mutations';
import { MessagesScrollWrapper, HasNextPage, NextPageButton } from './style';

class MessagesWithData extends Component {
  state: {
    subscription: ?Object,
  };

  state = {
    subscription: null,
  };

  componentDidMount() {
    this.props.forceScrollToBottom();
    this.subscribe();
  }

  componentDidUpdate(prev) {
    const { contextualScrollToBottom, data } = this.props;

    if (
      prev.data.networkStatus === 1 &&
      prev.data.loading &&
      !this.props.data.loading
    ) {
      this.subscribe();
      setTimeout(() => this.props.forceScrollToBottom());
    }
    // force scroll to bottom when a message is sent in the same thread
    if (prev.data.messages !== data.messages && contextualScrollToBottom) {
      contextualScrollToBottom();
    }
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  subscribe = () => {
    this.setState({
      subscription: this.props.subscribeToNewMessages(),
    });
  };

  unsubscribe = () => {
    const { subscription } = this.state;
    if (subscription) {
      // This unsubscribes the subscription
      subscription();
    }
  };

  render() {
    const {
      data: { error, loading, messages, hasNextPage, fetchMore, networkStatus },
    } = this.props;
    const { subscription } = this.state;

    if (error) {
      return <div>Error!</div>;
    }

    // NOTE(@mxstbr): The networkStatus check shouldn't be there, but if I remove
    // it the loading indicator doesn't show when switching between threads which
    // is hella annoying as the old msgs stick around until the new ones are there.
    // TODO: FIXME and remove the networkStatus === 7
    if (messages && networkStatus === 7) {
      let sortedMessages = sortAndGroupMessages(messages);

      return (
        <MessagesScrollWrapper>
          {hasNextPage &&
            <HasNextPage>
              <NextPageButton
                loading={networkStatus === 3}
                onClick={() => fetchMore()}
              >
                {networkStatus === 3
                  ? <Spinner size={16} color={'brand.default'} />
                  : 'Load previous messages'}
              </NextPageButton>
            </HasNextPage>}
          <ChatMessages
            toggleReaction={this.props.toggleReaction}
            messages={sortedMessages}
            forceScrollToBottom={this.props.forceScrollToBottom}
            contextualScrollToBottom={this.props.contextualScrollToBottom}
            threadId={this.props.id}
            threadType={'directMessageThread'}
          />
        </MessagesScrollWrapper>
      );
    }

    if (networkStatus === 7) {
      return null;
    } else {
      return <Loading />;
    }
  }
}

const Messages = compose(
  toggleReactionMutation,
  getDirectMessageThreadMessages,
  pure
)(MessagesWithData);

export default Messages;
