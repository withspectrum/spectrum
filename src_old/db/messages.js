//@flow
//$FlowFixMe
import database from 'firebase/database';
import { getStory } from './stories';
import { getMessageGroup } from './messageGroups';
import { getActiveMessageGroupObject } from '../actions/messageGroups';
import { createNotifications } from './notifications';
import { getUserInfo } from './users';
import { ACTIVITY_TYPES } from './types';

const UNIQUE = (v, i, a) => a.indexOf(v) === i;

type MessageTypes = 'messages' | 'messages_private';
type LocationTypes = 'stories' | 'message_groups';

export const getMessageKey = (type: MessageTypes): string =>
  database().ref(type).push().key;

type GetMessageReturn = Promise<Object>;
export const getMessage = (
  type: MessageTypes,
  key: string
): GetMessageReturn => {
  const db = database();

  return db
    .ref(`${type}/${key}`)
    .once('value')
    .then(snapshot => snapshot.val());
};

type GetMessagesFromLocationReturn = Promise<Array<Object>>;
export const getMessagesFromLocation = (
  location: LocationTypes,
  key: string
): GetMessagesFromLocationReturn => {
  const getDataFunc = location === 'stories'
    ? getStory
    : location === 'message_groups' ? getMessageGroup : getStory;

  return getDataFunc(key).then(data => {
    if (!data.messages) return Promise.resolve([]);
    const messages = Object.keys(data.messages);
    const type = location === 'stories'
      ? 'messages'
      : location === 'message_groups' ? 'messages_private' : 'messages';
    return Promise.all(messages.map(key => getMessage(type, key)));
  });

  return Promise.resolve([]);
};

type CreateMessageProps = {
  storyId: string,
  frequencyId: string,
  user: Object,
  message: {
    type: string,
    content: string,
  },
  key: ?string,
};
type CreateMessageReturn = Promise<Object>;
export const createMessage = ({
  storyId,
  frequencyId,
  user,
  message,
  key,
}: CreateMessageProps): CreateMessageReturn => {
  const db = database();
  const messageType = 'messages';
  const locationType = 'stories';
  const id = key || getMessageKey(messageType);

  return db
    .ref()
    .update({
      [`${messageType}/${id}`]: {
        id,
        storyId,
        frequencyId,
        timestamp: database.ServerValue.TIMESTAMP,
        userId: user.uid,
        message,
      },
      [`${locationType}/${storyId}/last_activity`]: database.ServerValue
        .TIMESTAMP,
      [`${locationType}/${storyId}/participants/${user.uid}`]: {
        id: user.uid,
        last_activity: database.ServerValue.TIMESTAMP,
      },
      [`${locationType}/${storyId}/messages/${id}`]: {
        id,
      },
    })
    .then(() => db.ref(`${locationType}/${storyId}`).once('value'))
    .then(snapshot => snapshot.val())
    .then(story => {
      // If this is a new story after the intro of story.participants, use that rather than
      // getting them manually
      if (story.participants) return Object.keys(story.participants);

      // If this is an old story without story.participants, get all the messages
      // and manually get all users that sent messages
      return getMessagesFromLocation(locationType, storyId).then(messages =>
        messages
          // - Everybody who's sent a message in that story before
          .map(({ userId }) => userId)
          // - Creator of story
          .concat([story.creator.uid])
          .filter(UNIQUE)
      );
    })
    .then(users =>
      createNotifications({
        // Avoid notifying the sender
        users: users.filter(uid => uid !== user.uid),
        activityType: ACTIVITY_TYPES.NEW_MESSAGE,
        ids: {
          frequency: frequencyId,
          story: storyId,
        },
        sender: {
          uid: user.uid,
          displayName: user.displayName,
          photoURL: user.photoURL,
        },
        content: message.type === 'text' ? message.content.substr(0, 140) : '',
      })
    )
    .then(() => db.ref(`${messageType}/${id}`).once('value'))
    .then(snapshot => snapshot.val());
};

type CreatePrivateMessageProps = {
  messageGroupId: string,
  userId: string,
  message: {
    type: string,
    content: string,
  },
  key: ?string,
};
type CreatePrivateMessageReturn = Promise<Object>;
export const createPrivateMessage = ({
  messageGroupId,
  userId,
  message,
  key,
}: CreatePrivateMessageProps): CreatePrivateMessageReturn => {
  const db = database();
  const messageType = 'messages_private';
  const locationType = 'message_groups';
  const id = key || getMessageKey(messageType);

  return db
    .ref()
    .update({
      [`${messageType}/${id}`]: {
        id,
        messageGroupId,
        timestamp: database.ServerValue.TIMESTAMP,
        userId,
        message,
      },
      [`${locationType}/${messageGroupId}/last_activity`]: database.ServerValue
        .TIMESTAMP,
      [`${locationType}/${messageGroupId}/users/${userId}/last_activity`]: database
        .ServerValue.TIMESTAMP,
      [`${locationType}/${messageGroupId}/messages/${id}`]: {
        id,
      },
      [`${locationType}/${messageGroupId}/snippet`]: message.type === 'text'
        ? message.content
        : 'Sent an attachment', // save a snippet of the last text message, or assume a media message
    })
    .then(() => db.ref(`${messageType}/${id}`).once('value'))
    .then(snapshot => snapshot.val());
};

export const setLastActivityOnMessageGroupParticipants = (
  messageGroup: Object
) => {
  const users = Object.keys(messageGroup.users);
  Promise.all(
    users.map(user => setLastActiveOnMessageGroupForUser(messageGroup.id, user))
  );
};

export const setLastActiveOnMessageGroupForUser = (
  messageGroup: string,
  user: string
) => {
  const db = database();
  return db.ref(`/users/${user}/messageGroups/${messageGroup}`).update({
    ['last_activity']: database.ServerValue.TIMESTAMP,
  });
};

type CreateReactionProps = {
  type: MessageTypes,
  messageId: string,
  uid: string,
};
export const createReaction = ({
  type,
  messageId,
  uid,
}: CreateReactionProps) => {
  const db = database();

  const TYPE = 'like';

  return db
    .ref()
    .update({
      [`${type}/${messageId}/reactions/${uid}`]: {
        type: TYPE,
        timestamp: database.ServerValue.TIMESTAMP,
      },
    })
    .then(() => Promise.all([getMessage(type, messageId), getUserInfo(uid)]))
    .then(([message, user]) => {
      // no notifications needed for private message reactions
      if (type === 'messages_private') return;
      createNotifications({
        users: [message.userId],
        activityType: ACTIVITY_TYPES.REACTION,
        ids: {
          frequency: message.frequencyId,
          story: message.storyId,
        },
        sender: {
          uid,
          displayName: user.displayName,
          photoURL: user.photoURL,
        },
        content: TYPE,
      });
    });
};

type DeleteReactionProps = {
  type: MessageTypes,
  messageId: string,
  uid: string,
};
export const deleteReaction = ({
  type,
  messageId,
  uid,
}: DeleteReactionProps) => {
  const db = database();

  return db.ref(`/${type}/${messageId}/reactions/${uid}`).remove();
};
