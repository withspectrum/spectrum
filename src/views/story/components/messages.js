// @flow
import React from 'react';
// $FlowFixMe
import compose from 'recompose/compose';
//$FlowFixMe
import lifecycle from 'recompose/lifecycle';
import { sortAndGroupMessages } from '../../../helpers/messages';
import ChatMessages from '../../../components/chatMessages';
import { displayLoadingCard } from '../../../components/loading';
import { getStoryMessages } from '../queries';
import { toggleReactionMutation } from '../mutations';

const lifecycles = lifecycle({
  state: {
    subscribed: false,
  },
  componentDidUpdate() {
    if (!this.props.loading && !this.state.subscribed) {
      this.setState({
        subscribed: true,
      });
      this.props.subscribeToNewMessages();
    }
  },
});

const MessagesWithData = ({ data, toggleReaction }) => {
  if (data.error) {
    return <div>Error!</div>;
  }

  if (!data.story && !data.story.messageConnection) {
    return <div>No messages yet!</div>;
  }

  const sortedMessages = sortAndGroupMessages(
    data.story.messageConnection.edges
  );
  return (
    <ChatMessages toggleReaction={toggleReaction} messages={sortedMessages} />
  );
};

const Messages = compose(
  toggleReactionMutation,
  getStoryMessages,
  lifecycles,
  displayLoadingCard
)(MessagesWithData);

export default Messages;
