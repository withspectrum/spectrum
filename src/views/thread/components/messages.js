// @flow
import React, { Component } from 'react';
// $FlowFixMe
import compose from 'recompose/compose';
import { sortAndGroupMessages } from '../../../helpers/messages';
import ChatMessages from '../../../components/chatMessages';
import Icon from '../../../components/icons';
import { HorizontalRule } from '../../../components/globals';
import { LoadingChat } from '../../../components/loading';
import { Button } from '../../../components/buttons';
import { NullState } from '../../../components/upsell';
import { ChatWrapper } from '../style';
import { getThreadMessages } from '../queries';
import { toggleReactionMutation } from '../mutations';

export const EmptyChat = () =>
  <ChatWrapper>
    <HorizontalRule>
      <hr />
      <Icon glyph={'message'} />
      <hr />
    </HorizontalRule>
    <NullState
      heading={`ðŸ”¥ This thread is hot off the presses...`}
      copy={`Why don't you kick off the conversation?`}
    />
  </ChatWrapper>;

class MessagesWithData extends Component {
  state: {
    subscription: ?Object,
  };

  state = {
    subscription: null,
  };

  componentDidUpdate(prevProps) {
    // force scroll to bottom when a message is sent in the same thread
    if (
      prevProps &&
      prevProps.data &&
      prevProps.data.thread &&
      prevProps.data.thread.messageConnection !==
        this.props.data.thread.messageConnection
    ) {
      this.props.contextualScrollToBottom();
    }
  }

  componentDidMount() {
    const { currentUser } = this.props;
    if (!currentUser || !currentUser.id) return;

    this.subscribe();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  subscribe = () => {
    this.setState({
      subscription: this.props.subscribeToNewMessages(),
    });
  };

  unsubscribe = () => {
    const { subscription } = this.state;
    if (subscription) {
      // This unsubscribes the subscription
      subscription();
    }
  };

  render() {
    const {
      data: { networkStatus, error },
      data,
      currentUser,
      toggleReaction,
      forceScrollToBottom,
    } = this.props;
    const dataExists = data.thread && data.thread.messageConnection;
    const messagesExist =
      dataExists && data.thread.messageConnection.edges.length > 0;
    if (networkStatus === 8 || error) {
      return (
        <NullState
          heading="Sorry, we lost connection to the server..."
          copy="Mind reloading the page?"
        >
          <Button
            icon="view-reload"
            onClick={() => window.location.reload(true)}
          >
            Reload
          </Button>
        </NullState>
      );
    }

    if (messagesExist) {
      const unsortedMessages = data.thread.messageConnection.edges.map(
        message => message.node
      );
      const sortedMessages = sortAndGroupMessages(unsortedMessages);

      return (
        <ChatWrapper>
          <HorizontalRule>
            <hr />
            <Icon glyph={'message'} />
            <hr />
          </HorizontalRule>
          <ChatMessages
            threadId={data.thread.id}
            toggleReaction={toggleReaction}
            messages={sortedMessages}
            threadType={'story'}
            forceScrollToBottom={forceScrollToBottom}
          />
        </ChatWrapper>
      );
    }

    if (networkStatus === 7) {
      if (currentUser) {
        return <EmptyChat />;
      } else {
        return null;
      }
    } else {
      return (
        <ChatWrapper>
          <HorizontalRule>
            <hr />
            <Icon glyph={'message'} />
            <hr />
          </HorizontalRule>
          <LoadingChat />
        </ChatWrapper>
      );
    }
  }
}

const Messages = compose(toggleReactionMutation, getThreadMessages)(
  MessagesWithData
);

export default Messages;
