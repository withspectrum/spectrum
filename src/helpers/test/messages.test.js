// @flow
import { sortAndGroupMessages } from '../messages';

// TODO: FIXME
type Message = Object;

// 01.01.2017
const FIRST_JAN = 1483225200000;

const createMessage = ({
  timestamp,
  body,
  senderId,
}: {
  timestamp?: Date,
  body?: string,
  senderId?: string,
}): Message => ({
  id: 'whatever',
  timestamp: timestamp || new Date(FIRST_JAN),
  content: {
    body: body || 'Hey',
  },
  sender: {
    id: senderId || 'asdf123',
  },
  messageType: 'text',
});

const filterRobo = messageGroups =>
  messageGroups.filter(group => group[0].sender.id !== 'robo');

it('should sort messages by timestamp', () => {
  const one = createMessage({
    timestamp: new Date(FIRST_JAN - 10000),
    body: 'First',
  });
  const two = createMessage({
    timestamp: new Date(FIRST_JAN),
    body: 'Second',
  });
  const messages = [two, one];

  expect(filterRobo(sortAndGroupMessages(messages))).toEqual([[one, two]]);
});

it('should group messages by sender', () => {
  const one = createMessage({
    senderId: 'first',
  });
  const two = createMessage({
    senderId: 'first',
  });
  const three = createMessage({
    senderId: 'second',
  });
  const four = createMessage({
    senderId: 'first',
  });

  const messages = [one, two, three, four];

  expect(filterRobo(sortAndGroupMessages(messages))).toEqual([
    [one, two],
    [three],
    [four],
  ]);
});

it('should add a timestamp above the first message', () => {
  const timestamp = new Date(FIRST_JAN);
  const messages = [createMessage({ timestamp })];

  expect(sortAndGroupMessages(messages)).toMatchSnapshot();
});

const SEVEN_HOURS = 25200000;
it("should add a timestamp between two messages if there's more than six hours between them", () => {
  const first = createMessage({
    timestamp: new Date(FIRST_JAN),
  });
  const second = createMessage({
    timestamp: new Date(FIRST_JAN + SEVEN_HOURS),
  });

  const messages = [first, second];

  const result = sortAndGroupMessages(messages);

  // Should have three message groups, two robo texts + two groups
  expect(result.length).toEqual(4);
  // Expect a robo text timestamp to be between the message groups
  expect(result[2][0].sender.id).toEqual('robo');
});

describe('lastSeen', () => {
  it('should insert a last seen timestamp between two messages from different users', () => {
    const first = createMessage({
      timestamp: new Date(FIRST_JAN),
      senderId: 'first',
    });
    const second = createMessage({
      // Second one is 5000ms after first one
      timestamp: new Date(FIRST_JAN + 5000),
      senderId: 'second',
    });

    const messages = [first, second];

    // lastSeen is 2500 after first one
    const result = sortAndGroupMessages(messages, FIRST_JAN + 2500);

    // Between the two messages should be an unseen timestamp
    expect(result[2][0].sender.id).toEqual('robo');
    expect(result[2][0].message.type).toEqual('unseen-messages-below');
  });

  // NOTE(@mxstbr): This is a bug, is fixed in #2192, can remove the .skip here after that's merged
  it.skip(
    'should only insert one last seen robotext between messages from different users',
    () => {
      const first = createMessage({
        timestamp: new Date(FIRST_JAN),
        senderId: 'first',
      });
      const second = createMessage({
        timestamp: new Date(FIRST_JAN + 5000),
        senderId: 'second',
      });
      const third = createMessage({
        timestamp: new Date(FIRST_JAN + 6000),
        senderId: 'second',
      });

      const messages = [first, second, third];

      const result = sortAndGroupMessages(messages, FIRST_JAN + 2500);

      // Even though there's two unseen messages there should only be one robotext
      const robotexts = result.filter(
        group =>
          group[0].sender.id === 'robo' &&
          group[0].message.type === 'unseen-messages-below'
      );
      expect(robotexts.length).toEqual(1);
    }
  );

  it('should insert a last seen robotext between messages from the same user', () => {
    const first = createMessage({
      timestamp: new Date(FIRST_JAN),
    });
    const second = createMessage({
      // Second one is 5000ms after first one
      timestamp: new Date(FIRST_JAN + 5000),
    });

    const messages = [first, second];

    // lastSeen is 2500 after first one
    const result = sortAndGroupMessages(messages, FIRST_JAN + 2500);

    // Between the two messages should be an unseen timestamp
    expect(result[2][0].sender.id).toEqual('robo');
    expect(result[2][0].message.type).toEqual('unseen-messages-below');
  });

  it('should only insert one last seen robotext between messages from the same user', () => {
    const first = createMessage({
      timestamp: new Date(FIRST_JAN),
    });
    const second = createMessage({
      // Second one is 5000ms after first one
      timestamp: new Date(FIRST_JAN + 5000),
    });
    const third = createMessage({
      timestamp: new Date(FIRST_JAN + 6000),
    });

    const messages = [first, second, third];

    // lastSeen is 2500 after first one
    const result = sortAndGroupMessages(messages, FIRST_JAN + 2500);

    // Even though there's two unseen messages there should only be one robotext
    const robotexts = result.filter(
      group =>
        group[0].sender.id === 'robo' &&
        group[0].message.type === 'unseen-messages-below'
    );
    expect(robotexts.length).toEqual(1);
  });
});
