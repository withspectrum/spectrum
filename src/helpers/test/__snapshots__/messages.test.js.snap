// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a timestamp above the first message 1`] = `
Array [
  Array [
    Object {
      "message": Object {
        "content": 2016-12-31T23:00:00.000Z,
        "type": "timestamp",
      },
      "sender": Object {
        "id": "robo",
      },
      "timestamp": 2016-12-31T23:00:00.000Z,
    },
  ],
  Array [
    Object {
      "content": Object {
        "body": "Hey",
      },
      "id": "whatever",
      "messageType": "text",
      "sender": Object {
        "id": "asdf123",
      },
      "timestamp": 2016-12-31T23:00:00.000Z,
    },
  ],
]
`;
